### ISSUE TITLE
Add a hasZeroData selector for Search Console

### ISSUE THREAD

**Issue by tofumatt** [google/site-kit-wp#4672]
## Feature Description

For the new Zero States features to be properly implemented, we need a way to explicitly check if a module has "zero data" as opposed to "gathering data". Obviously any site that's gathering data will also have "zero data", but we can use logic tailored to each "module" (eg Google Service) that Site Kit uses to check if there's no data available from the Google API because the site is new or if there's data because there's been nothing of note.

Essentially, we want to be able to tell the difference in Search Console: "Do I have no data in Search Console because my site is very new or is there is no data because my site is not popular?"

---------------

_Do not alter or remove anything below. The following sections will be managed by moderators only._

## Acceptance criteria

* There should be a `hasZeroData()` selector for the Search Console module's datastore that checks to see if there is no data for Search Console despite `isGatheringData()` returning `false`.
* This selector should `return false` if any data is found. The selector should `return true` if no data for a "representative" Search Console request—likely whatever the default dashboard request would be (eg. 28 days)—is found but `isGatheringData() === false`.

## Implementation Brief

- In `assets/js/modules/search-console/datastore/report.js`:
    - Add a selector `hasZeroData` that checks for zero data state as described in the AC
    - The `hasZeroData` selector will have the following logic:
        - Get the report data from `select( MODULES_SEARCH_CONSOLE ).getReport( args )` selector
        - Reuse the logic for getting the `args` to pass to `getReport( args )`  from `isGatheringData`, see:
        https://github.com/google/site-kit-wp/blob/27b8e040ec85ab9fb94efebebf44defc544a26d3/assets/js/modules/search-console/datastore/report.js#L137-L160
        - Return `undefined` if the `report` data is `undefined`
        - Get the gathering data state from the `isGatheringData()` selector
        - Get whether the report data has zero state or not, by using the `isZeroReport( report )` helper function from `assets/js/modules/search-console/util/is-zero-report.js`
        - Return `undefined` if either `isGatheringData() === undefined || getReport() === undefined`
        - Return `false` if `isGatheringData() === false && isZeroReport() === false`
        - Return `true` otherwise (essentially, if `isGatheringData() === true || isZeroReport() === true`). Note, this can be the default return value, to ensure we always return from the function

### Test Coverage

- Add test cases for the new selector `hasZeroData`

## QA Brief

**QA:Eng**
- Verify the changes met the Acceptance Criteria 
- Run the newly added tests for `hasZeroData` from `assets/js/modules/search-console/datastore/report.test.js`
- As an extra verification, could call the `hasZeroData` selector from `assets/js/modules/search-console/components/dashboard/SearchFunnelWidget/index.js` component to check what `hasZeroData` returns
- For the above, verify if the report data is available, it should return `false` otherwise `true`

## Changelog entry

* N/A

**Comment by tofumatt** [comment]
The `isGatheringData() === false && isZeroReport()` is a bit dangerous because if `isGatheringData()` is `undefined` and `isZeroReport()` is `true`, we could return `true` while `isGatheringData()` resolves. Once it does it could then return `true`, which isn't ideal.

We should be clear in the logic for this selector that both `isGatheringData()` and the `getReport()` selectors much be resolved (or rather, must not return `undefined`) before we can return anything other than `undefined` ourselves. It's possible that this is an edge case given `isGatheringData()` will fetch reports too but they might not be the same... I think it's good to explicitly protect against returning anything until we have all the data.

---

A separate note: this IB is a bit too bullet-point in its display of the intended logic. I find it a bit hard-to-parse in bullet form, and it ends up reading a bit more like pseudo-code.

For instance, writing `select( MODULES_SEARCH_CONSOLE ).getReport` but then showing the assembly of the `args` value _underneath_ the `getReport` approach makes it hard to parse the logic because it isn't being written in the order the function would approach things. eg. We'd need to assemble the report `args` _before_ we call `getReport`. Could you try to reformat the logic so it's a bit less like pseudo-code?

Hopefully that makes sense, sorry if it doesn't! 😅

**Comment by hussain-t** [comment]
Thanks for the clarifications @tofumatt 👍 I have updated the IB.

**Comment by tofumatt** [comment]
Looks good, just a few clarification questions here:

> Reuse the logic from `isGatheringData`, see:

What does this point mean/refer to? Do you mean reuse the logic used to get the `args` passed to `getReport`?

> Return `true` if `isGatheringData()` is not `undefined` but resolved and returns `false` AND `isZeroReport()` is not `undefined` but resolved and returns `true`

This sentence is tough-to-parse. Let's rephrase, because I think it means:

`hasZeroData` should:
  * return `undefined` if either `isGatheringData() === undefined || getReport() === undefined`
  * return `false if `isGatheringData() === false && isZeroReport() === false`
  * return `true` otherwise (essentially, if `isGatheringData() === true || isZeroReport() === true`) (note this can be the default return value, to ensure we always return from the function

That is the intended logic, so the IB should reflect it.

**Comment by tofumatt** [comment]
Looks good, thanks! 👍🏻 

IB ✅

**Comment by asvinb** [comment]
### QA: ✅ 
- AC ☑️ 
- Tests for `hasZeroData` ☑️ 
- Calling `hasZeroData` returns `true` or `false` ☑️
